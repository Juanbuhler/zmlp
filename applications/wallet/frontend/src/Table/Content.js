import PropTypes from 'prop-types'
import { useRouter } from 'next/router'
import useSWR from 'swr'

import { colors, constants, spacing, typography } from '../Styles'

import Pagination from '../Pagination'

import GearSvg from '../Icons/gear.svg'

const SIZE = 20

const TableContent = ({
  url,
  columns,
  expandColumn,
  renderEmpty,
  renderRow,
}) => {
  const {
    query: { page = 1 },
  } = useRouter()

  const parsedPage = parseInt(page, 10)
  const from = parsedPage * SIZE - SIZE

  const { data: { count = 0, results } = {}, error, revalidate } = useSWR(
    `${url}?from=${from}&size=${SIZE}`,
  )

  const hasException = !results || results.length === 0 || error

  if (hasException) return renderEmpty

  return (
    <div css={{ height: hasException ? '100%' : 'auto' }}>
      <table
        css={{
          width: '100%',
          height: hasException ? '100%' : 'auto',
          borderSpacing: 0,
          boxShadow: constants.boxShadows.default,
          whiteSpace: 'nowrap',
          tr: {
            backgroundColor: colors.structure.lead,
            '&:nth-of-type(2n)': {
              backgroundColor: colors.structure.mattGrey,
            },
            ':hover': {
              backgroundColor: colors.structure.iron,
              boxShadow: constants.boxShadows.tableRow,
              td: {
                border: constants.borders.tableRow,
                borderLeft: '0',
                borderRight: '0',
                '&:first-of-type': {
                  borderLeft: constants.borders.tableRow,
                },
                '&:last-of-type': {
                  borderRight: constants.borders.tableRow,
                },
              },
            },
          },
          td: {
            maxWidth: `calc(100vw / ${columns.length - 1})`,
            overflow: 'hidden',
            textOverflow: 'ellipsis',
            whiteSpace: 'nowrap',
            fontWeight: typography.weight.extraLight,
            color: colors.structure.pebble,
            padding: `${spacing.base}px ${spacing.normal}px`,
            border: constants.borders.transparent,
            borderLeft: '0',
            borderRight: '0',
            ':first-of-type': {
              borderLeft: constants.borders.transparent,
            },
            ':last-of-type': {
              borderRight: constants.borders.transparent,
              overflow: 'visible',
            },
          },
        }}>
        <thead>
          <tr>
            {columns.map(column => (
              <th
                key={column}
                css={{
                  textAlign: 'left',
                  fontSize: typography.size.regular,
                  lineHeight: typography.height.regular,
                  fontWeight: typography.weight.medium,
                  color: colors.structure.pebble,
                  backgroundColor: colors.structure.iron,
                  padding: `${spacing.moderate}px ${spacing.normal}px`,
                  borderBottom: constants.borders.default,
                  [`:nth-of-type(${expandColumn})`]: { width: '100%' },
                  '&:not(:last-child)': {
                    borderRight: constants.borders.default,
                  },
                }}>
                {column === '#Actions#' ? (
                  <div css={{ display: 'flex' }}>
                    <GearSvg width={20} />
                  </div>
                ) : (
                  column
                )}
              </th>
            ))}
          </tr>
        </thead>
        <tbody>
          {results.map(result => renderRow({ result, revalidate }))}
        </tbody>
      </table>

      <div>&nbsp;</div>

      {count > 0 && !error && (
        <Pagination
          currentPage={parsedPage}
          totalPages={Math.ceil(count / SIZE)}
        />
      )}

      <div>&nbsp;</div>
    </div>
  )
}

TableContent.propTypes = {
  url: PropTypes.string.isRequired,
  columns: PropTypes.arrayOf(PropTypes.string).isRequired,
  expandColumn: PropTypes.number.isRequired,
  renderEmpty: PropTypes.node.isRequired,
  renderRow: PropTypes.func.isRequired,
}

export default TableContent
