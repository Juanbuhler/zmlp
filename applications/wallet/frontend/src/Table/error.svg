<svg width="604" height="201" xmlns="http://www.w3.org/2000/svg"><g fill="none" fill-rule="evenodd"><path d="M330.321802 163.687501c-.755049-1.573055-2.619204-2.134731-4.336455-1.307034-1.7168.827924-2.44274 2.638045-1.688368 4.210874l11.376284 23.720464c.47185.984008 1.36974 1.595676 2.4635 1.678694.083945.006334.168115.009501.252511.009501 1.193952 0 2.406183-.627955 3.07819-1.614903.618301-.908002.702245-1.992899.230622-2.976228l-11.376284-23.721368zM352.273599 149.816685c-.741032-.735128-1.740206-.989721-2.741092-.699476-1.164418.338146-2.145183 1.371174-2.440783 2.570431-.254074 1.030841.025472 2.043529.766718 2.778656l17.872533 17.722461.000214.000218c.548176.543308 1.236552.811025 1.943123.811025.821728 0 1.667429-.362205 2.344033-1.075023 1.258171-1.325242 1.313395-3.210634.128215-4.385613l-17.872961-17.722679zM277.785323 166.135923c-1.615132-.483399-3.140578.358248-3.54833 1.95811l-6.137899 24.118769c-.25477 1.000128-.012897 2.000471.663514 2.744923.60498.665749 1.488144 1.042275 2.357616 1.042275.262112 0 .523033-.034191.774629-.105152.933562-.262988 1.614934-.987441 1.869704-1.987785l6.138693-24.119199c.407156-1.598786-.50339-3.168757-2.117927-3.651941M306.027147 171.683727c-.196297-1.718911-1.850667-2.862386-3.848618-2.660715-1.996998.202119-3.355613 1.650567-3.159792 3.369254l2.954216 25.923042c.122299 1.074908.798751 1.944559 1.85614 2.385546.482296.200999 1.01456.299146 1.549202.299146.829446 0 1.664126-.236179 2.325112-.690388.934376-.642211 1.400017-1.627262 1.277242-2.702618l-2.953502-25.923267zM267.653031 42.1488013C268.116123 43.3210579 269.216188 44 270.465645 44c.44105 0 .900579-.0845367 1.357214-.2619975 1.750397-.6796042 2.597988-2.3879956 1.971702-3.9734445l-9.443951-23.9135058c-.391402-.9919264-1.229642-1.65299-2.299428-1.8138966-1.245226-.1867312-2.609343.3423404-3.395263 1.3177127-.67549.8383036-.83824 1.8878386-.446393 2.8793236l9.443505 23.9146094zM247.729852 57.1879313c.548291.5438585 1.23724.8120687 1.944173.8120687.821258 0 1.666708-.3620619 2.34324-1.0741533 1.259291-1.3255183 1.314741-3.211959.129312-4.3878397l-17.876495-17.7265924-.000428-.0002187c-1.185643-1.1752244-3.028978-1.0616835-4.287413.2638348-1.258863 1.3255183-1.314098 3.2115215-.128669 4.3876209l17.87628 17.7252797zM321.156794 39.8552281c.311069.0980226.618818.1447719.916217.1447719 1.218694 0 2.258323-.787628 2.590872-2.072694l6.232307-24.104736c.258736-.9994.02851-2.0039704-.631315-2.7560518-.768005-.8755253-2.004078-1.26459976-3.074951-.9683775-.920709.2546434-1.597131.9739788-1.855477 1.9735942l-6.233088 24.1049514c-.413391 1.5978767.470411 3.1793804 2.055435 3.6785418M294.466888 36.9984378c2.063504-.0502142 3.581794-1.3859116 3.531918-3.1074771l-.758411-25.95716405c-.031051-1.076369-.647657-1.99026721-1.692123-2.50758493-1.215366-.60212389-2.823873-.56306842-4.002564.09685759-1.012927.56730873-1.57599 1.50999641-1.544451 2.58681176l.757434 25.95761043C290.808078 35.7555809 292.34226 37 294.348065 37c.039363 0 .079215-.0006695.118823-.0015622" fill="#4A4A4A"/><path d="M239.247802 141.452511c-12.883346 3.038124-25.83391-4.981825-28.869085-17.877693l-.475401-2.020538c-3.035174-12.895868 4.976785-25.85902 17.860131-28.8971443l21.652169-5.1058215 11.484153 48.7953748-21.651967 5.105822zm18.272375-67.0141611l-32.254362 7.6060844c-17.050263 4.0205447-28.229396 19.9986477-26.768873 36.9750267-3.177734.702309-6.099185 1.378803-8.794817 2.003056-13.91355 3.22229-22.31176 5.16721-34.304618 3.583913-7.941897-1.048484-15.694933-3.619079-22.420811-7.433505-7.171427-4.067094-12.998085-9.365619-17.317918-15.748404-2.318316-3.4251569-4.226215-7.1712825-6.246212-11.1371467-1.920897-3.7717381-3.907183-7.6717417-6.317289-11.3672523-9.6422533-14.7836684-25.8302558-25.4457652-44.4137431-29.2524663-8.2881419-1.6977395-17.1191061-2.1016438-26.2477797-1.2005328-3.0280666.2988119-6.09553.7431676-9.1902056 1.3269688l.0040615-.0058949c-4.0970588.7478429-8.2452929 1.7701048-12.415256 3.0580452-.224196.0691129-.4398627.1361931-.6500464.2020537-.0280245.0087408-.056049.0176848-.0840735.0264256-.25668806.0804962-.50301928.15835-.73939978.2337644-.81088258.2563278-1.62217131.518957-2.43427235.7935794l.00284307.0085375c-1.40650457.4817579-2.0449756.7378824-2.0449756.7378824-3.62815631 1.4568605-5.69059645 4.7480603-4.5830198 7.3137772 1.10757666 2.565717 4.98409495 3.4778048 8.61468818 2.0270425 0 0 .40188746-.160586 1.25968058-.4598045l.0006092.0016262c16.6497974-5.6304701 32.2848239-7.0121195 46.4715121-4.1061228C72.5916326 62.891404 86.4225316 71.949267 94.5928894 84.4761941c2.1132092 3.2401781 3.8913429 6.7313989 5.7736546 10.427316 2.071782 4.0679074 4.214031 8.2742439 6.885902 12.2216099 5.204432 7.689833 12.175423 14.045583 20.719035 18.890808 7.85031 4.452094 16.875008 7.449157 26.098315 8.666968 3.419598.451471 6.579664.655151 9.620118.655151 9.244427-.000204 17.378434-1.883735 28.301081-4.413473 2.640801-.611446 5.501127-1.273914 8.599255-1.959149 5.576468 17.035531 23.408574 27.28478 41.155388 23.099787L274 144.459331l-16.479823-70.0209811zM375.609958 108.80275l-21.678088 5.099786-11.4979-48.7376963 21.678088-5.0997861c12.898769-3.0345331 25.864836 4.9759359 28.90344 17.8565606l.476173 2.0181493c3.038808 12.8806246-4.982945 25.8284535-17.881713 28.8629865m227.32318-75.7185657c-1.980125-1.9373421-6.168092-1.7014176-9.347596.5112374l-.000204-.000406c-13.764297 7.4076636-31.3758-2.8343425-40.648575-9.5985942-1.988461-1.4504686-4.032428-2.9967689-6.0093-4.492311C534.890361 10.3976687 522.443571.98139831 506.842895.09251928c-8.175259-.465758-16.607309.83324796-24.383656 3.75590983-8.363328 3.14356191-15.752352 7.98996269-21.961718 14.40459129-6.269345 6.4761477-10.460565 13.0934034-14.513732 19.4928045-2.91275 4.5989034-5.924719 9.3541423-9.725974 14.0706019-10.721424 13.3017153-14.181712 14.0551713-30.809556 17.675455-1.087554.2367366-2.247693.4893098-3.470861.7589378-6.372834-15.6333478-23.408944-24.7751174-40.366165-20.7859225l-32.293176 7.5970935 2.808447 11.9040353-30.793494 7.2444251c-3.009937.7079765-4.892672 3.7465703-4.183494 6.7524757l.09068.3847478c.709177 3.0057024 3.752051 4.8855862 6.762191 4.1776097l30.793494-7.2444251 5.544106 23.5000689-30.793698 7.244222c-3.009936.70818-4.892468 3.746774-4.183291 6.752476l.09068.384951c.709178 3.005702 3.752051 4.885586 6.761988 4.177407l30.793697-7.244222L345.817607 127l32.293176-7.597094c17.876427-4.205425 29.309265-21.50872 26.510374-39.3384802 1.046483-.2300365 2.048643-.4485002 2.992857-.6539697 8.814697-1.9192722 14.638164-3.187214 20.076747-6.0810452 5.389177-2.8676399 10.011637-7.1146869 16.486334-15.1478956 4.164382-5.166584 7.333518-10.1703354 10.39835-15.009224 3.953134-6.2412383 7.687089-12.1365082 13.231195-17.8636667 5.15536-5.3255589 11.289904-9.3492694 18.233862-11.9592603 6.459246-2.4278701 13.453017-3.50739726 20.22334-3.1216343 12.509006.7126463 23.701113 9.1797368 34.524807 17.3680628 2.011436 1.5217333 4.091391 3.0954431 6.15142 4.5980913 8.287491 6.0453113 16.428185 10.2188603 24.196195 12.4045119 7.629754 2.1466693 14.681471 2.3001623 21.049629.4698187l.002033.0036546c5.646172-1.6220317 8.769154-4.0596474 8.769154-4.0596474 3.083741-2.4067547 3.973059-5.9744528 1.976058-7.9280376" fill="gray"/></g></svg>