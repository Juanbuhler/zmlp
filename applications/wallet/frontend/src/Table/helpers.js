import Router from 'next/router'

export const onRowClickRouterPush = (...args) => (event) => {
  const { target: { localName } = {} } = event || {}

  if (['a', 'button', 'svg', 'path'].includes(localName)) return

  Router.push(...args)
}
