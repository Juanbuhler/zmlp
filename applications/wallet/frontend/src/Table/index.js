/* eslint-disable react/jsx-props-no-spreading */
import PropTypes from 'prop-types'
import Router, { useRouter } from 'next/router'

import { constants, spacing } from '../Styles'

import { getQueryString } from '../Fetch/helpers'

import InputSearch, { VARIANTS as INPUT_SEARCH_VARIANTS } from '../Input/Search'
import SuspenseBoundary, { ROLES } from '../SuspenseBoundary'

import TableContent from './Content'

const Table = ({ role, searchLabel, ...props }) => {
  const {
    pathname,
    query,
    query: { search = '' },
  } = useRouter()

  return (
    <div css={{ height: '100%', display: 'flex', flexDirection: 'column' }}>
      {!!searchLabel && (
        <div
          css={{
            display: 'flex',
            alignItems: 'flex-end',
            justifyContent: 'space-between',
            paddingBottom: spacing.normal,
            flexShrink: 0,
          }}
        >
          <div
            css={{
              width: constants.form.maxWidth,
              paddingRight: spacing.spacious,
            }}
          >
            <InputSearch
              aria-label={`Filter ${searchLabel}`}
              placeholder={`Filter ${searchLabel}`}
              value={search}
              onChange={({ value }) => {
                const queryParamFilter = getQueryString({
                  query: query.query,
                  sort: query.sort,
                  search: value,
                })
                const href = `${pathname}${queryParamFilter}`
                const as = href
                  .split('/')
                  .map((s) =>
                    s.replace(/\[(.*)\]/gi, (_, group) => query[group]),
                  )
                  .join('/')
                Router.push(href, as)
              }}
              variant={INPUT_SEARCH_VARIANTS.EXTRADARK}
            />
          </div>
        </div>
      )}
      <SuspenseBoundary role={role}>
        <TableContent {...props} />
      </SuspenseBoundary>
    </div>
  )
}

Table.defaultProps = {
  role: null,
  searchLabel: '',
}

Table.propTypes = {
  role: PropTypes.oneOf(Object.keys(ROLES)),
  searchLabel: PropTypes.string,
}

export { Table as default, ROLES }
