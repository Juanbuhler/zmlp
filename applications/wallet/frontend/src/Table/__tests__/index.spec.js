import TestRenderer from 'react-test-renderer'

import Table from '..'

const noop = () => () => {}

jest.mock('../../Fetch/Ahead', () => 'FetchAhead')

describe('<Table />', () => {
  it('should render properly without data', () => {
    require('swr').__setMockUseSWRResponse({ data: {} })

    const component = TestRenderer.create(
      <Table
        legend="Stuff"
        url=""
        refreshKeys={[]}
        refreshButton={false}
        columns={['ColumnOne, ColumnTwo']}
        expandColumn={2}
        renderEmpty="Empty"
        renderRow={noop}
      />,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should prefetch previous and next page', () => {
    require('next/router').__setUseRouter({
      query: { page: 2 },
    })

    require('swr').__setMockUseSWRResponse({
      data: { count: 100, results: [] },
    })

    const component = TestRenderer.create(
      <Table
        legend="Stuff"
        url="/api/v1/stuff"
        refreshKeys={[]}
        refreshButton={false}
        columns={['ColumnOne, ColumnTwo']}
        expandColumn={2}
        renderEmpty="Empty"
        renderRow={noop}
      />,
    )

    expect(component.toJSON()).toMatchSnapshot()
  })
})
