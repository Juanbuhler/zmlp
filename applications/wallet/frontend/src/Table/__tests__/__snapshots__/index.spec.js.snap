// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> should prefetch previous and next page 1`] = `
.emotion-0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding-bottom: 16px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-2 {
  color: #b3b3b3;
  font-size: 15px;
  line-height: 19px;
  font-weight: 400;
}

.emotion-3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
}

.emotion-5 {
  height: auto;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.5);
  white-space: nowrap;
}

.emotion-5 tr {
  background-color: #252525;
}

.emotion-5 tr:nth-of-type(2n) {
  background-color: #2e2e2e;
}

.emotion-5 tr:hover {
  background-color: #4a4a4a;
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.5);
}

.emotion-5 tr:hover .actions {
  color: #b3b3b3;
}

.emotion-5 tr:hover td {
  border: 1px solid #808080;
  border-left: 0;
  border-right: 0;
}

.emotion-5 tr:hover td:first-of-type {
  border-left: 1px solid #808080;
}

.emotion-5 tr:hover td:last-of-type {
  border-right: 1px solid #808080;
}

.emotion-5 td {
  max-width: calc(100vw / 0);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #e0e0e0;
  padding: 8px 16px;
  border: 1px solid transparent;
  border-left: 0;
  border-right: 0;
}

.emotion-5 td:first-of-type {
  border-left: 1px solid transparent;
}

.emotion-5 td:last-of-type {
  border-right: 1px solid transparent;
  overflow: visible;
}

.emotion-6 {
  text-align: left;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  color: #e0e0e0;
  background-color: #4a4a4a;
  padding: 12px 16px;
  border-bottom: 1px solid #2e2e2e;
  width: auto;
}

.emotion-6:nth-of-type(2) {
  width: 100%;
}

.emotion-6:not(:last-child) {
  border-right: 1px solid #2e2e2e;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-8 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.emotion-9 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-10 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border-radius: 2px;
  padding: 8px 16px;
  white-space: nowrap;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  color: #ffffff;
  background-color: transparent;
  padding-left: 64px;
  padding-right: 64px;
}

.emotion-10,
.emotion-10:hover,
.emotion-10:visited,
.emotion-10 .noop {
  background-color: #808080;
}

.emotion-10:hover,
.emotion-10 .noop {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #b3b3b3;
}

.emotion-10[aria-disabled=true],
.emotion-10 .noop {
  color: #2e2e2e;
  background-color: #808080;
}

.emotion-11 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-12 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border-radius: 2px;
  padding: 12px;
  white-space: nowrap;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  color: #ffffff;
  background-color: transparent;
  margin-right: 8px;
}

.emotion-12,
.emotion-12:hover,
.emotion-12:visited,
.emotion-12 .noop {
  background-color: #808080;
}

.emotion-12:hover,
.emotion-12 .noop {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #b3b3b3;
}

.emotion-12[aria-disabled=true],
.emotion-12 .noop {
  color: #2e2e2e;
  background-color: #808080;
}

.emotion-13 {
  -webkit-transform: rotate(90deg);
  -moz-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-14 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border-radius: 2px;
  padding: 12px;
  white-space: nowrap;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  color: #ffffff;
  background-color: transparent;
}

.emotion-14,
.emotion-14:hover,
.emotion-14:visited,
.emotion-14 .noop {
  background-color: #808080;
}

.emotion-14:hover,
.emotion-14 .noop {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #b3b3b3;
}

.emotion-14[aria-disabled=true],
.emotion-14 .noop {
  color: #2e2e2e;
  background-color: #808080;
}

.emotion-15 {
  -webkit-transform: rotate(90deg);
  -moz-transform: rotate(90deg);
  -ms-transform: rotate(90deg);
  transform: rotate(90deg);
}

.emotion-16 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-17 {
  padding: 8px;
  margin-left: 8px;
  text-align: center;
  width: 2.5ch;
  box-sizing: content-box;
  border-radius: 2px;
  color: #ffffff;
  background-color: #3b3b3b;
  border: 2px solid transparent;
  -moz-appearance: textfield;
}

.emotion-17:hover {
  border: 2px solid #ffffff;
}

.emotion-17:focus {
  border: 2px solid #3A5DFF;
  outline: #3A5DFF;
  color: #000000;
  background-color: #ffffff;
}

.emotion-17::-webkit-outer-spin-button,
.emotion-17::-webkit-inner-spin-button {
  -webkit-appearance: none;
  margin: 0;
}

.emotion-18 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding-left: 8px;
  padding-right: 8px;
  color: #b3b3b3;
}

.emotion-20 {
  -webkit-transform: rotate(-90deg);
  -moz-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

.emotion-21 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  border-radius: 2px;
  padding: 12px;
  white-space: nowrap;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  border: none;
  cursor: pointer;
  color: #ffffff;
  background-color: transparent;
  margin-left: 8px;
}

.emotion-21,
.emotion-21:hover,
.emotion-21:visited,
.emotion-21 .noop {
  background-color: #808080;
}

.emotion-21:hover,
.emotion-21 .noop {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #b3b3b3;
}

.emotion-21[aria-disabled=true],
.emotion-21 .noop {
  color: #2e2e2e;
  background-color: #808080;
}

.emotion-22 {
  -webkit-transform: rotate(-90deg);
  -moz-transform: rotate(-90deg);
  -ms-transform: rotate(-90deg);
  transform: rotate(-90deg);
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <h3
      className="emotion-2"
    >
      Number of 
      Stuff
      : 
      100
    </h3>
  </div>
  <div
    className="emotion-3"
  >
    <div
      className="emotion-4"
    >
      <table
        className="emotion-5"
      >
        <thead>
          <tr>
            <th
              className="emotion-6"
            >
              ColumnOne, ColumnTwo
            </th>
          </tr>
        </thead>
        <tbody />
      </table>
    </div>
    <div>
      Â 
    </div>
    <div
      className="emotion-7"
    >
      <div
        className="emotion-8"
      />
      <div
        className="emotion-9"
      >
        <a
          className="emotion-10"
          href="/undefined?page=3"
          onClick={[Function]}
          onMouseEnter={[Function]}
          rel="next"
        >
          Next
        </a>
      </div>
      <div
        className="emotion-11"
      >
        <a
          aria-label="First page"
          className="emotion-12"
          href="/undefined"
          onClick={[Function]}
          onMouseEnter={[Function]}
          rel="prev"
          title="First page"
        >
          <svg
            className="emotion-13"
            data-file-name="doubleChevron.svg"
            height={16}
          />
        </a>
        <a
          aria-label="Previous page"
          className="emotion-14"
          href="/undefined"
          onClick={[Function]}
          onMouseEnter={[Function]}
          rel="prev"
          title="Previous page"
        >
          <svg
            className="emotion-15"
            data-file-name="chevron.svg"
            height={16}
          />
        </a>
        <form
          action=""
          className="emotion-16"
          method="get"
          onSubmit={[Function]}
        >
          <input
            aria-label="Go to page"
            className="emotion-17"
            max={5}
            min="1"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            type="number"
            value="2"
          />
        </form>
        <div
          className="emotion-18"
        >
          of 
          5
        </div>
        <a
          aria-label="Next page"
          className="emotion-14"
          href="/undefined?page=3"
          onClick={[Function]}
          onMouseEnter={[Function]}
          rel="next"
          title="Next page"
        >
          <svg
            className="emotion-20"
            data-file-name="chevron.svg"
            height={16}
          />
        </a>
        <a
          aria-label="Last page"
          className="emotion-21"
          href="/undefined?page=5"
          onClick={[Function]}
          onMouseEnter={[Function]}
          rel="next"
          title="Last page"
        >
          <svg
            className="emotion-22"
            data-file-name="doubleChevron.svg"
            height={16}
          />
        </a>
      </div>
    </div>
    <FetchAhead
      url="/api/v1/stuff?from=40&size=20"
    />
    <FetchAhead
      url="/api/v1/stuff?from=0&size=20"
    />
    <div>
      Â 
    </div>
  </div>
</div>
`;

exports[`<Table /> should render properly without data 1`] = `
.emotion-0 {
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.emotion-1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-end;
  -webkit-box-align: flex-end;
  -ms-flex-align: flex-end;
  align-items: flex-end;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  padding-bottom: 16px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
}

.emotion-2 {
  color: #b3b3b3;
  font-size: 15px;
  line-height: 19px;
  font-weight: 400;
}

.emotion-3 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  position: relative;
}

.emotion-4 {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
}

.emotion-5 {
  height: 100%;
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.5);
  white-space: nowrap;
}

.emotion-5 tr {
  background-color: #252525;
}

.emotion-5 tr:nth-of-type(2n) {
  background-color: #2e2e2e;
}

.emotion-5 tr:hover {
  background-color: #4a4a4a;
  box-shadow: 0 0 5px 0 rgba(0, 0, 0, 0.5);
}

.emotion-5 tr:hover .actions {
  color: #b3b3b3;
}

.emotion-5 tr:hover td {
  border: 1px solid #808080;
  border-left: 0;
  border-right: 0;
}

.emotion-5 tr:hover td:first-of-type {
  border-left: 1px solid #808080;
}

.emotion-5 tr:hover td:last-of-type {
  border-right: 1px solid #808080;
}

.emotion-5 td {
  max-width: calc(100vw / 0);
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  color: #e0e0e0;
  padding: 8px 16px;
  border: 1px solid transparent;
  border-left: 0;
  border-right: 0;
}

.emotion-5 td:first-of-type {
  border-left: 1px solid transparent;
}

.emotion-5 td:last-of-type {
  border-right: 1px solid transparent;
  overflow: visible;
}

.emotion-6 {
  text-align: left;
  font-size: 15px;
  line-height: 19px;
  font-weight: 500;
  color: #e0e0e0;
  background-color: #4a4a4a;
  padding: 12px 16px;
  border-bottom: 1px solid #2e2e2e;
  width: auto;
}

.emotion-6:nth-of-type(2) {
  width: 100%;
}

.emotion-6:not(:last-child) {
  border-right: 1px solid #2e2e2e;
}

.emotion-7 {
  height: 100%;
  pointer-events: none;
}

.emotion-8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  text-align: center;
  color: #808080;
  font-size: 15px;
  line-height: 19px;
}

<div
  className="emotion-0"
>
  <div
    className="emotion-1"
  >
    <h3
      className="emotion-2"
    >
      Number of 
      Stuff
      : 
      0
    </h3>
  </div>
  <div
    className="emotion-3"
  >
    <div
      className="emotion-4"
    >
      <table
        className="emotion-5"
      >
        <thead>
          <tr>
            <th
              className="emotion-6"
            >
              ColumnOne, ColumnTwo
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            className="emotion-7"
          >
            <td
              colSpan={1}
            >
              <div
                className="emotion-8"
              >
                Empty
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</div>
`;
